tests/handlers/test_password_providers.py

```diff
@@ -719,6 +721,25 @@ def custom_auth_no_local_user_fallback_test_body(self):
         channel = self._send_password_login("localuser", "localpass")
         self.assertEqual(channel.code, 400, channel.result)
 
+    def test_on_logged_out(self):
+        """Tests that the on_logged_out callback is called when the user logs out."""
+        self.register_user("rin", "password")
+        tok = self.login("rin", "password")
+
+        on_logged_out = Mock(return_value=make_awaitable(None))
+        self.hs.get_password_auth_provider().on_logged_out_callbacks.append(
+            on_logged_out
+        )
+
+        channel = self.make_request(
+            "POST",
+            "/_matrix/client/v3/logout",
+            {},
+            access_token=tok,
+        )
+        self.assertEqual(channel.code, 200)
+        on_logged_out.assert_called_once()
```

Comment:
Hold on, does this actually detect the bug? I.e. does it fail without the fix?

`on_logged_out` is a substitution for `callback`. The problem isn't that `callback` didn't get called, it's that `callback(...)` didn't get awaited.

Reply:
Oh that's true huh

Reply:
So it turns out there's some magic in our test rig that turns `RuntimeWarning: coroutine 'PasswordAuthProviderTests.test_on_logged_out.<locals>.on_logged_out' was never awaited` into an error and fails the test. I've improved the test to check that we actually run the code in the function (and so that we await on it) just in case.

Reply:
Interesting. But we still didn't see a failure--- even with that magic. Does that mean that the tests simply never hit this path?

Reply:
Yes, the test suite was missing a test that checked that `on_logged_out` gets called.

Reply:
Oh right, because the test suite never added any callbacks to run in the first place. Gotcha, thanks.